hera_librarian/async_transfers/globus.py

```diff
@@ -0,0 +1,271 @@
+        transfers. For the Confidential App, the secret is the Client Secret
+        generated when the app was created.
+
+        Once the authenticator is created, the way they work downstream is
+        effectively interchangeable. Note that Globus as a service will perform
+        further checking to see if the user/app has permission to read and write
+        to specific endpoints. We will do our best to handle this as it comes up
+        to provide the user with nicer error messages, though we may not have
+        caught all possible failure modes.
+        """
+        if self.authorizer is None:
+            if self.native_app:
+                try:
+                    client = globus_sdk.NativeAppAuthClient(self.client_id)
+                    self.authorizer = globus_sdk.RefreshTokenAuthorizer(
+                        self.secret, client
+                    )
+                except globus_sdk.AuthAPIError as e:
+                    return False
+            else:
+                try:
+                    client = globus_sdk.ConfidentialAppAuthClient(
+                        self.client_id, self.secret
+                    )
+                    tokens = client.oauth2_client_credentials_tokens()
+                    transfer_tokens_info = tokens.by_resource_server[
+                        "transfer.api.globus.org"
+                    ]
+                    transfer_token = transfer_tokens_info["access_token"]
+                    self.authorizer = globus_sdk.AccessTokenAuthorizer(transfer_token)
+                except globus_sdk.AuthAPIError as e:
+                    return False
+
+        return True
+
+    @property
+    def valid(self) -> bool:
+        """
+        This is easy to check if we're using a system with a Globus Connect
+        Personal (GCP) endpoint, but harder to verify for environments with
+        Globus Connect Server endpoints (e.g., NERSC). For now, we're lazy and
+        assume we can always use Globus (though this is obviously NOT always
+        true).
+        """
+        return True
+
+    def _get_transfer_data(self, local_endpoint, remote_endpoint, label):
+        """
+        This is a helper function to create a TransferData object, which is needed
+        both for single-book transfers and batch transfers.
+        """
+        # create a TransferData object that contains options for the transfer
+        transfer_data = globus_sdk.TransferData(
+            source_endpoint=local_endpoint,
+            destination_endpoint=remote_endpoint,
+            label=label,
+            sync_level="exists",
+            verify_checksum=False,  # we do this ourselves
+            preserve_timestamp=True,
+            notify_on_succeeded=False,
+        )
+
+        return transfer_data
+
+    def transfer(
+        self,
+        local_path: Path,
+        remote_path: Path,
+        local_endpoint: str,
+        remote_endpoint: str,
```

Comment:
How do you intend for this to work? This doesn't implement the abstract base type.